src/hardware/disk_noise.cpp

```diff
@@ -0,0 +1,479 @@
+			active_samples++;
+		}
+	}
+
+	// Fill the output buffer with silence if no active samples
+	if (active_samples == 0) {
+		mix_channel_->AddSamples_sfloat(frames, out.data());
+		return;
+	}
+
+	const float mix_scale = 1.0f / static_cast<float>(active_devices_.size());
+
+	auto scale_sample = [](float sample, float volume, float scale) -> float {
+		return static_cast<float>(sample) * volume * scale;
+	};
+
+	for (auto* device : active_devices_) {
+		for (int i = 0; i < frames; ++i) {
+			float mixed_l = 0.0f;
+			float mixed_r = 0.0f;
+
+			if (!device->spin_up_sample_.empty() &&
+			    device->spin_up_pos_ + 1 <
+			            device->spin_up_sample_.size()) {
+				mixed_l += scale_sample(
+				        device->spin_up_sample_[device->spin_up_pos_],
+				        device->spin_volume_,
+				        mix_scale);
+				mixed_r += scale_sample(
+				        device->spin_up_sample_[device->spin_up_pos_++],
+				        device->spin_volume_,
+				        mix_scale);
+			} else if (!device->spin_sample_.empty() &&
+			           (device->spin_pos_ + 1 <
+			                    device->spin_sample_.size() ||
+			            device->loop_spin_sample_)) {
+				mixed_l += scale_sample(
+				        device->spin_sample_[device->spin_pos_],
+				        device->spin_volume_,
+				        mix_scale);
+				mixed_r += scale_sample(
+				        device->spin_sample_[device->spin_pos_++],
+				        device->spin_volume_,
+				        mix_scale);
+
+				// Loop the spin sound if enabled. Used for
+				// persistent HDD noise Not used for floppy
+				// noise because motor should stop after r/w
+				// operations aredone
+				if (device->spin_pos_ >= device->spin_sample_.size() &&
+				    device->loop_spin_sample_) {
+					device->spin_pos_ = 0;
+				}
+			}
+
+			if (!device->current_seek_sample_.empty() &&
+			    device->seek_pos_ + 1 <
+			            device->current_seek_sample_.size()) {
+				mixed_l += scale_sample(
+				        device->current_seek_sample_[device->seek_pos_],
+				        device->seek_volume_,
+				        mix_scale);
+				mixed_r += scale_sample(
+				        device->current_seek_sample_[device->seek_pos_++],
+				        device->seek_volume_,
+				        mix_scale);
+			}
+
+			out[i * 2] += mixed_l;
+			out[i * 2 + 1] += mixed_r;
+		}
+
+		if (!device->current_seek_sample_.empty() &&
+		    device->seek_pos_ >= device->current_seek_sample_.size()) {
+			device->current_seek_sample_.clear();
+		}
+	}
+
+	mix_channel_->AddSamples_sfloat(frames, out.data());
+}
+
+void DiskNoiseDevice::Shutdown()
+{
+	std::lock_guard<std::mutex> lock(device_mutex_);
+	active_devices_.erase(std::remove(active_devices_.begin(),
+	                                  active_devices_.end(),
+	                                  this),
+	                      active_devices_.end());
+
+	if (active_devices_.empty() && mix_channel_) {
+		mix_channel_->Enable(false);
+		MIXER_DeregisterChannel(mix_channel_);
+		mix_channel_.reset();
+	}
```

Comment:
Yeah, this will be simpler if `DiskNoises` manages the single channel, not all the individual devices.

Reply:
Yep, very nice 👍🏻